c/tskit/trees.c

```diff
@@ -2124,6 +2124,580 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+        tsk_bit_array_add(&allele_samples_row, &state_row);
+
+        /* Get the index for the alternate allele that we must subtract from */
+        alt_allele = site->ancestral_state;
+        alt_allele_length = site->ancestral_state_length;
+        if (mutation.parent != TSK_NULL) {
+            parent_mut = site->mutations[mutation.parent - site->mutations[0].id];
+            alt_allele = parent_mut.derived_state;
+            alt_allele_length = parent_mut.derived_state_length;
+        }
+        for (allele = 0; allele < num_alleles; allele++) {
+            if (alt_allele_length == allele_lengths[allele]
+                && tsk_memcmp(alt_allele, alleles[allele], allele_lengths[allele])
+                       == 0) {
+                break;
+            }
+        }
+        tsk_bug_assert(allele < num_alleles);
+
+        tsk_bit_array_get_row(out_allele_samples, allele, &alt_allele_samples_row);
+        tsk_bit_array_subtract(&alt_allele_samples_row, &allele_samples_row);
+    }
+    *out_num_alleles = num_alleles;
+out:
+    tsk_safe_free(alleles);
+    tsk_safe_free(allele_lengths);
+    return ret;
+}
+
+static int
+get_mutation_samples(const tsk_tree_t *tree, const tsk_size_t *site_offsets,
+    tsk_bit_array_t *allele_samples, tsk_size_t **num_alleles)
+{
+    int ret = 0;
+
+    const tsk_treeseq_t *restrict ts = tree->tree_sequence;
+
+    const tsk_size_t num_samples = tsk_treeseq_get_num_samples(ts);
+    const tsk_site_t *restrict tree_sites = ts->tree_sites[tree->index];
+    const tsk_flags_t *restrict flags = ts->tables->nodes.flags;
+    const tsk_id_t *restrict mut_nodes = ts->tables->mutations.node;
+
+    const tsk_site_t *restrict site;
+
+    tsk_bit_array_t mut_samples, mut_samples_row, out_row;
+    tsk_size_t num_mutations, mut_samples_offset, mut_offset;
+    tsk_size_t s, m, n;
+
+    tsk_id_t node;
+    tsk_size_t num_nodes;
+    tsk_id_t *nodes = tsk_malloc(tsk_tree_get_size_bound(tree) * sizeof(*nodes));
+
+    tsk_memset(&mut_samples, 0, sizeof(mut_samples));
+
+    if (nodes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    // Get the number of mutations
+    num_mutations = 0;
+    for (s = 0; s < tree->sites_length; s++) {
+        num_mutations += tree_sites[s].mutations_length;
+    }
+
+    ret = tsk_bit_array_init(&mut_samples, num_samples, num_mutations);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Current position minus num ancestral alleles to this point, if there are any sites
+    mut_offset = tree->sites_length
+                     ? (site_offsets[tree_sites[0].id] - (tsk_size_t) tree_sites[0].id)
+                     : 0;
+
+    // Traverse down the tree, recording all samples below each mutation.
+    // We currently perform one preorder traversal per mutation, but there
+    // might be more clever ways to do this. The preorder traversals begin at
+    // the focal mutation's node.
+    for (m = 0; m < num_mutations; m++) {
```

Comment:
I don't get why you need to use the arrays here: why not use the site and mutation objects within the tree struct?
```
for (s = 0; s < tree->sites_length; s++) {
     site = tree->sites[s];
     for (m = 0; m < site.mutations_length; m++) {
           node = site.mutations[m].node;
           // do traversal rooted at node. 
    }
}

Reply:
See my comment about offsets, I've cleaned things up, using the suggested loop.